ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -332,41 +352,25 @@ typedef struct {
 static int update_object_record_entry(st_data_t *key, st_data_t *value, st_data_t data, int existing) {
   object_record_update_data *update_data = (object_record_update_data*) data;
   if (existing) {
-    // Object was already tracked. One of 3 things could have happened:
-    //
-    // 1. Ruby did some smart memory re-use.
-    //    TODO: Research this case and document it.
-    //
-    // 2. We missed a free.
-    //    TODO: Research this case and document it.
-    //
-    // 3. An unknown thing happened.
-    //
-    // Except for #3 where we're kinda screwed anyway, dropping the existing object record is
-    // acceptable behaviour and is equivalent to treating this allocation as a combined
-    // free+allocation.
-    VALUE obj = (VALUE) (*key);
-    object_record *existing_record = (object_record*) (*value);
-    commit_free(update_data->heap_recorder, obj, existing_record);
+    rb_raise(rb_eRuntimeError, "Object ids are supposed to be unique. We got 2 allocation recordings with the same id");
```

Comment:
I've mentioned it on slack, but this will break for Ruby <= 2.6. I think for now if needed we can disable heap profiling for such Rubies -- I think it makes sense to have a good implementation for modern Rubies, and we can later do a pass on older ones if there's enough interest. (E.g. rather than trying to do a one size fits all)